### PR TITLE
[feat] add ddp that works with oss with reduce() not all_reduce()

### PR DESCRIPTION
**Pull request by min-xu-ai**
This is still a WIP:

* ~~add ddp code from fairseq as-is, no changes~~
* ~~make the change for ddp & oss & reduce()~~
* ~~add unit test~~
* ~~clean up mypy, style, etc~~

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by min-xu-ai** [comment]
Hi guys, is this something reasonable for the first cut? Someone may have to take it from here for bigger changes since I need to help data loading and VISSL before end of Aug. Sorry about that!

File: requirements-test.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@ flake8 == 3.7.9
 isort == 4.3.21
 mypy == 0.770
 pytest == 5.4.1
+pytest-cov == 2.10.0
```
Comment (blefaudeux) [inline requirements-test.txt:6]: nice, good idea !

File: tests/nn/data_parallel/test_oss_ddp.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    model = Sequential(Linear(2, 3), Linear(3, 4)).to(device)
+    optimizer = OSS(model.parameters(), lr=0.1, momentum=0.99)
+    ddp = OssDdp(model, optimizer, world_size)
+    input_tensor = torch.rand((64, 2)).to(device)
+    output = ddp(input_tensor).sum()
+    output.backward()
+    ddp.reduce()
+    optimizer.step()
+    # TODO: I need to figure out a way to verify the grads are reduced correctly
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_oss_ddp.py:49]: could you specify who is "I" when this is written down ? gives more context when somebody chimes in and would like a POC

File: tests/optim/test_oss.py [hunk 3]
```diff
@@ -234,6 +235,7 @@ def closure():
     optimizer.load_state_dict(optimizer_state_dict)
 
 
+@skip_if_not_enough_gpu
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:238]: is that strictly needed ? we can have several processes storing data to the same gpu, no ? I'm a bit surprised that this is needed actually

File: tests/nn/data_parallel/test_oss_ddp.py (hunk 2 above)
Comment (msbaines) [reply tests/nn/data_parallel/test_oss_ddp.py:49]: TODO(name) is recommended format

**Comment by min-xu-ai** [comment]
> It's nice to have one thesis per change.
> 
> Could you put the code coverage stuff into a separate change.

Of course. I will make a separate PR on that.

File: requirements-test.txt (hunk 1 above)
Comment (min-xu-ai) [reply requirements-test.txt:6]: Thanks!

File: tests/nn/data_parallel/test_oss_ddp.py (hunk 2 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_oss_ddp.py:49]: Of course. Done.

File: tests/optim/test_oss.py (hunk 3 above)
Comment (min-xu-ai) [reply tests/optim/test_oss.py:238]: It appears to be needed. It fails on my system with only 2 GPUs. But I didn't debug deeply into it. Here is more info:

```
>       raise Exception(msg)
E       Exception:
E
E       -- Process 2 terminated with the following error:
E       Traceback (most recent call last):
E         File "/scratch/m1n/env1/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 20, in _wrap
E           fn(i, *args)
E         File "/private/home/m1n/git/fairscale/tests/optim/test_oss.py", line 197, in run_test_collect_shards
E           target = torch.rand((batch, target_width), device=device)
E       RuntimeError: CUDA error: invalid device ordinal

/scratch/m1n/env1/lib/python3.6/site-packages/torch/multiprocessing/spawn.py:119: Exception
```

When I print the `device` param above, here is what I see:

```
tests/optim/test_oss.py::test_collect_shards cuda:1
cuda:2
cuda:0
FAILED
```

It seems to hardcode to 3 GPUs?

Comment (msbaines) [inline tests/optim/test_oss.py:238]: This should be a separate change. Each change should have a single thesis.

Correct fix in this case might be to set world_size to number of cuda devices or 3.

Comment (min-xu-ai) [reply tests/optim/test_oss.py:238]: You are right. Thanks for the suggestion. Let me move this out.

File: fairscale/optim/oss.py [hunk 4]
```diff
@@ -27,13 +27,13 @@ class OSS(Optimizer):
 
     .. _ZeRO: https://arxiv.org/abs/1910.02054
 
-    Pipe combines pipeline parallelism with checkpointing to reduce peak
-    memory required to train while minimizing device under-utilization.
+    We use a greedy algorithm to pack a number of parameters
```
Comment (min-xu-ai) [inline fairscale/optim/oss.py:30]: I will move this doc change out to a separate PR as well.

**Pull request merged**